File: dynamice programming/1239.cpp
class Solution {
public:
    bool hasDuplicate(string &s1, string &s2){
        int arr[26] = {0};
        // check also this: {'aa', 'bb'}
        for(char &ch: s1){
            if(arr[ch-'a'] > 0)
                return true;
            arr[ch-'a']++;
        }
        for(char &ch: s2){
            if(arr[ch-'a'] > 0)
                return true;
        }
    return false; // no duplicate found at all
    }
    unordered_map<string, int> mp;
    int solve(int i, vector<string>& arr, string temp, int n){
        if(i >= n)
            return temp.length();
        int include = 0; 
        int exclude = 0;
        if(mp.find(temp) != mp.end()){
            return mp[temp];
        }
        if(hasDuplicate(arr[i], temp)){ // exclude only
            exclude = solve(i + 1, arr, temp, n);
        }
        else{ // include and exclude
            exclude = solve(i+1, arr, temp, n);
            include = solve(i+1, arr, temp+arr[i], n);
        }
        return mp[temp]=max(include, exclude);
    }
    int maxLength(vector<string>& arr) {
        string temp = ""; // no concatenation till now
        mp.clear();
        int n = arr.size();
        int i = 0;
        return solve(i, arr, temp, n);
    }
};

File: dynamice programming/1239.py
# Microsoft, Apple
# only concatenate when two string doesn't have any common char or element
# Recursion + Memoization
# Tc = O(n * 2^n)
# Sc = O(n)

class Solution:
    def hasDuplicate(self, s1, s2):
        arr = [0] * 26
        for ch in s1:
            if arr[ord(ch) - ord('a')] > 0:
                return True
            arr[ord(ch) - ord('a')] += 1
        for ch in s2:
            if arr[ord(ch) - ord('a')] > 0:
                return True
        return False  # no duplicate found at all

    def solve(self, i, arr, temp, n, memo):
        if i >= n:
            return len(temp)
        if temp in memo:
            return memo[temp]

        include = 0
        exclude = 0

        if self.hasDuplicate(arr[i], temp):  # exclude only
            exclude = self.solve(i + 1, arr, temp, n, memo)
        else:  # include and exclude
            exclude = self.solve(i + 1, arr, temp, n, memo)
            include = self.solve(i + 1, arr, temp + arr[i], n, memo)

        memo[temp] = max(include, exclude)
        return memo[temp]

    def maxLength(self, arr):
        memo = {}
        n = len(arr)
        return self.solve(0, arr, "", n, memo)
